### PR TITLE
feat(lep): recurring snapshot delete design

### PR DESCRIPTION
**Pull request by c3y1huang**
https://github.com/longhorn/longhorn/issues/3836

### CODE REVIEW

**Review by shuo-wu — COMMENTED** [review]
It seems that action `create` is responsible to create snapshots/backups only while action `delete` is responsible to delete snapshots/backups only. Then does it mean that users need to launch one `create` job + one `delete` job to achieve the effect of the previous recurring jobs?

What I am thinking is:
1. Add a new boolean field to indicate if `retain` should count the snapshots/backups not created by the current job.
2. Add a new boolean field to indicate that a job needs to create any snapshots/backups.

BTW, we plan to rely on recurring jobs to trigger snapshot purge periodically. Will we do anything about this? For example, if `retain == 0` or the current snapshot/backup count is not greater than `retain`, the job will blindly send a purge request only. I don't know if there is a separate ticket for it. [The ticket I found](https://github.com/longhorn/longhorn/issues/2732) is actually similar to what you mentioned.

**Comment by c3y1huang** [comment]
> It seems that action `create` is responsible to create snapshots/backups only while action `delete` is responsible to delete snapshots/backups only. Then does it mean that users need to launch one `create` job + one `delete` job to achieve the effect of the previous recurring jobs?

There is no difference between the new create action and the current recurring job implementation: periodically create snapshot/backup + recurring job owning snapshot purge.

And the delete action periodically purges all snapshots (except the latest backup).

> 
> What I am thinking is:
> 
> 1. Add a new boolean field to indicate if `retain` should count the snapshots/backups not created by the current job.

Do you mean we should respect the current job retain?

> 2. Add a new boolean field to indicate that a job needs to create any snapshots/backups.

I prefer to deal by action. It can also work if there are more recurring job actions to come, ex: https://github.com/longhorn/longhorn/issues/5186.

> 
> BTW, we plan to rely on recurring jobs to trigger snapshot purge periodically. Will we do anything about this? For example, if `retain == 0` or the current snapshot/backup count is not greater than `retain`, the job will blindly send a purge request only. I don't know if there is a separate ticket for it. [The ticket I found](https://github.com/longhorn/longhorn/issues/2732) is actually similar to what you mentioned.

Currently, recurring job cleanup by the `cleanupSnapshotNames` then follows with the purge. The plan is RecurringJob with delete action will have its own `cleanupSnapshotNames` so it can be used for purge.

**Comment by shuo-wu** [comment]
> There is no difference between the new create action and the current recurring job implementation:

OK. The action name `create` is a little bit misleading...

> Do you mean we should respect the current job retain?
I prefer to deal by action. It can also work if there are more recurring job actions to come

In your current design, when the action is `delete`, `retain` will calculate all kinds of snaps including user created or other jobs created. But when the action is `create`, `retain` just counts its own snaps, right?

> The plan is RecurringJob with delete action will have its own cleanupSnapshotNames so it can be used for purge.

For this part, the requirement is, having a mechanism triggers the snapshot purge only (without deleting existing valid snapshots). Since some operations like trim may generate `removed` or system snapshots without a following purge request, or somehow the previous purge/snapshot deletion may be interrupted, having this mechanism can help a lot. Can we achieve this here?

**Comment by c3y1huang** [comment]
> In your current design, when the action is `delete`, `retain` will calculate all kinds of snaps including user created or other jobs created. But when the action is `create`, `retain` just counts its own snaps, right?

right.

> For this part, the requirement is, having a mechanism triggers the snapshot purge only (without deleting existing valid snapshots). Since some operations like trim may generate `removed` or system snapshots without a following purge request, or somehow the previous purge/snapshot deletion may be interrupted, having this mechanism can help a lot. Can we achieve this here?

Sure, added action `purge`.

**Comment by shuo-wu** [comment]
> > In your current design, when the action is `delete`, `retain` will calculate all kinds of snaps including user created or other jobs created. But when the action is `create`, `retain` just counts its own snaps, right?
> 
> right.

OK. Here I am thinking if we can make it more intuitive... The naming of action `create` is misleading as well.


> > For this part, the requirement is, having a mechanism triggers the snapshot purge only (without deleting existing valid snapshots). Since some operations like trim may generate `removed` or system snapshots without a following purge request, or somehow the previous purge/snapshot deletion may be interrupted, having this mechanism can help a lot. Can we achieve this here?
> 
> Sure, added action `purge`.

Do we need one more action purge? How about asking users to set a `delete` job with a large `retain` number? I mean, a snap `delete` job will send a purge request regardless of the current snap count reaching the `retain` limit. Will this be tricky?

**Comment by c3y1huang** [comment]
> OK. Here I am thinking if we can make it more intuitive... The naming of action `create` is misleading as well.
 
 Any ideas?
 Currently:
 - action `create` recurringly creates snapshot/backup and deletes and purges its own snapshots.
 - action `delete` recurringly deletes and purges all snapshots.
 - action `purge` recurringly purges all snapshots.
 
> Do we need one more action purge? 

I think there could be two use cases:
1. user just wants to periodically clean up system snapshots.
1. user wants to periodically delete all snapshots.

> How about asking users to set a `delete` job with a large `retain` number? I mean, a snap `delete` job will send a purge request regardless of the current snap count reaching the `retain` limit. Will this be tricky?

Not sure if we should consider the `delete` job cleanup the snapshot of the `create` job before the `create` job reached the retaining number. Respecting other retaining is tricky and IMO over complexing the solution.

The assumption is the delete/purge action will not be set to a frequent interval as the snapshot/backup so I prefer to leave this to the user.

File: enhancements/20230103-recurring-snapshot-cleanup.md [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+### Related Issues
+
+https://github.com/longhorn/longhorn/issues/3836
+
+## Motivation
+
+### Goals
+
+Introduce recurring job actions (create, delete, purge).
```
Comment (innobead) [inline enhancements/20230103-recurring-snapshot-cleanup.md:16]: `create`, `delete`, `purge` are general. Sometimes generalization can help abstract similar behaviors from things, but on the other side, it will also limit what we can support. 

I would suggest just extending types of `RecurringJobType`, and it will be good enough and flexible for us. Then, we will have `Snapshot`, `Backup`, `SnapshotDelete`, `BackupDelete` or even `SystemBackup` in the future.

File: enhancements/20230103-recurring-snapshot-cleanup.md [hunk 2]
```diff
@@ -0,0 +1,134 @@
+# Recurring Snapshot Cleanup
+## Summary
+
+Longhorn recurring job automatically cleans up its owning expired snapshots during snapshot creation. For the non-recurring volume snapshots, the user needs to clean them manually.
```
Comment (innobead) [inline enhancements/20230103-recurring-snapshot-cleanup.md:4]: `Longhorn recurring job automatically cleans up its owning expired snapshots during snapshot creation.` should be `Longhorn recurring job can automatically clean up older snapshots of volumes to retain no more than the desired number of snapshot defined for each volume.`

Comment (c3y1huang) [reply enhancements/20230103-recurring-snapshot-cleanup.md:4]: This is a description of the current implementation. The suggested change doesn't look right.

I will revisit this and reword it to improve on clarity.

Comment (c3y1huang) [reply enhancements/20230103-recurring-snapshot-cleanup.md]: Update.

File: enhancements/20230103-recurring-snapshot-cleanup.md (hunk 1 above)
Comment (c3y1huang) [reply enhancements/20230103-recurring-snapshot-cleanup.md]: I see, updated. Thanks!

**Comment by shuo-wu** [comment]
>> How about asking users to set a delete job with a large retain number? I mean, a snap delete job will send a purge request regardless of the current snap count reaching the retain limit. Will this be tricky?
>
>Not sure if we should consider the delete job cleanup the snapshot of the create job before the create job reached the retaining number. Respecting other retaining is tricky and IMO over complexing the solution.

No. I mean, For `SnapshotDelete` jobs, it will blindly send a snapshot purge request. A purge request won't mark any valid snapshots as `Removed`, instead, it is just for cleaning up accumulated removed/system snaps. This does not involve respecting other retention.

File: enhancements/20230103-recurring-snapshot-cleanup.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+https://github.com/longhorn/longhorn/issues/3836
+
+## Motivation
+
+### Goals
+
+Introduce new recurring job types:
+- `snapshot-delete`: periodically deletes and purges snapshots.
+- `snapshot-purge`: periodically purges snapshots.
```
Comment (innobead) [inline enhancements/20230103-recurring-snapshot-cleanup.md:18]: When deleting a snapshot, it includes two steps `mark it as removed` and `purge`, so we should regard both of them as a deletion. If a snapshot has been marked remove, when deleting the snapshot, it will be just `purge`.

Basically, we don't need to disclose internal technical terms (delete or purge) to let users decide which deletion they should choose, but rather they should decide what types of snapshot they want to delete: system or all and the deletion will do the purge at the end.  

I would suggest two new job types: `snapshot-delete` and `system-snapshot-delete`. `snapshot-delete` will delete all types of snapshots including user created ones, but `system-snapshot-delete` will only delete system snapshots.

File: enhancements/20230103-recurring-snapshot-cleanup.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+   metadata:
+     name: recurring-snap-delete-per-min
+     namespace: longhorn-system
+   spec:
+     concurrency: 1
+     cron: '* * * * *'
+     groups: []
+     labels: {}
+     name: recurring-snap-delete-per-min
+     retain: 2
```
Comment (innobead) [inline enhancements/20230103-recurring-snapshot-cleanup.md:54]: Later, we will have different types of recurring jobs, so just need to emphasize the `retain` of the deletion jobs will supersede others. For example, if users have one snapshot recurring job to keep 5 retain, but also have a cleanup job with 2 retain. Eventually, there will be at most 2 snapshots retained. This is straight forward, but it can be briefly explained in the doc.

**Review by innobead — CHANGES_REQUESTED** [review]

File: enhancements/20230103-recurring-snapshot-cleanup.md (hunk 3 above)
Comment (innobead) [reply enhancements/20230103-recurring-snapshot-cleanup.md:18]: @shuo-wu WDYT? :point_up:

Comment (innobead) [reply enhancements/20230103-recurring-snapshot-cleanup.md]: > As I mentioned, I prefer to blindly send purge requests for all removing/system snapshots in `snapshot-delete` jobs.

Sounds good to me. WDYT? @c3y1huang

Comment (c3y1huang) [reply enhancements/20230103-recurring-snapshot-cleanup.md]: We can keep the snapshot purge only for the recurring cleanup if the other use case is not required. However, I think naming the task as `snapshot-delete` might be misleading since the task is for periodically deleting the system snapshots only. I prefer to use `system-snapshot-delete` as the task name. WDYT @shuo-wu ?

Comment (innobead) [reply enhancements/20230103-recurring-snapshot-cleanup.md]: All of you have fair points. 

How about using `snapshot-cleanup` instead? Cleanup just means the snapshots able to be cleaned up will be deleted accordingly, no matter what snapshot type each is, ex: removable or system snapshot.

Comment (c3y1huang) [reply enhancements/20230103-recurring-snapshot-cleanup.md]: Sounds good to me. WDYT @shuo-wu ?

Comment (shuo-wu) [reply enhancements/20230103-recurring-snapshot-cleanup.md]: OK. Let's do it this way. And the final consensus is below, right?
`snapshot-delete`: remove and purge all kinds of snapshots that exceed the retention count.
`snapshot-cleanup`: purge removed and system snapshots.

Comment (c3y1huang) [reply enhancements/20230103-recurring-snapshot-cleanup.md]: I thought the consensus is to have `snapshot-cleanup` only. Do we still want `snapshot-delete`?

Comment (c3y1huang) [reply enhancements/20230103-recurring-snapshot-cleanup.md]: After talking to @shuo-wu we still need the `snapshot-delete` to remove user-created snapshots.

So yes the consensus is below:
> snapshot-delete: remove and purge all kinds of snapshots that exceed the retention count.
> snapshot-cleanup: purge removed and system snapshots.

Comment (innobead) [reply enhancements/20230103-recurring-snapshot-cleanup.md]: After thinking, it should be ok to have snapshot-delete and snapshot-cleanup for giving users more options for snapshot cleanup strategy.

However what we really care is snapshot-cleanup.

File: enhancements/20230103-recurring-snapshot-cleanup.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+https://github.com/longhorn/longhorn/issues/3836
+
+## Motivation
+
+### Goals
+
+Introduce new recurring job types:
+- `snapshot-delete`: periodically deletes and purges snapshots.
+- `snapshot-cleanup`: periodically purges snapshots.
```
Comment (innobead) [inline enhancements/20230103-recurring-snapshot-cleanup.md:18]: snapshot-delete: periodically remove and purge all kinds of snapshots that exceed the retention count.
snapshot-cleanup: periodically purge to remove system snapshots.

File: enhancements/20230103-recurring-snapshot-cleanup.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+`None`
+
+## Proposal
+
+- Introduce two new `RecurringJobType`:
+  - snapshot-delete
+  - snapshot-cleanup
+- Recurring job periodically deletes and purges the snapshots for RecurringJob using the `snapshot-delete` task type. Longhorn will retain snapshots based on the given retain number.
+- Recurring job periodically purges the snapshots for RecurringJob using the `snapshot-cleanup` task type. Longhorn will retain system snapshots based on the given retain number.
+
```
Comment (innobead) [inline enhancements/20230103-recurring-snapshot-cleanup.md:31]: Not sure if `retain` is meaningful for `snapshot-delete`, because the target snapshots are removable and system snapshots. Can make it in general first.

File: enhancements/20230103-recurring-snapshot-cleanup.md [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+### User Stories
+
+- The user can create a RecurringJob with `spec.task=snapshot-delete` to instruct Longhorn periodically delete and purge snapshots.
+- The user can create a RecurringJob with `spec.task=snapshot-cleanup` to instruct Longhorn periodically purge snapshots.
```
Comment (innobead) [inline enhancements/20230103-recurring-snapshot-cleanup.md:35]: `The user can create a RecurringJob with `spec.task=snapshot-cleanup` to instruct Longhorn periodically purge removable or system snapshots.`

File: enhancements/20230103-recurring-snapshot-cleanup.md [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+     concurrency: 1
+     cron: '* * * * *'
+     groups: []
+     labels: {}
+     name: recurring-snap-delete-per-min
+     retain: 2
+     task: snapshot-delete
+   ```
+1. Assign the RecurringJob to volume.
+1. Longhorn deletes all expired snapshots. As a result of the above example, the user will see two snapshots after the job completes.
```
Comment (innobead) [inline enhancements/20230103-recurring-snapshot-cleanup.md:58]: Should the remaining snapshots be user-created, right? or the type does not matter but the creation time of snapshot is respected instead? (this question is more about the current behavior for retaining snapshots)

File: enhancements/20230103-recurring-snapshot-cleanup.md [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+1. Continue with the current implementation to purge snapshots.
+
+#### The RecurringJob `snapshot-cleanup` Task Type
+
+1. Do snapshot purge only in `doSnapshotCleanup`.
+
+### RecurringJob Mutate
+
+1. Mutate `Recurringjob.Spec.Retain` to 0.
+
```
Comment (innobead) [inline enhancements/20230103-recurring-snapshot-cleanup.md:100]: What's the section for? Do you mean some validation for `retain` in the mutation webhook?

**Review by innobead — CHANGES_REQUESTED** [review]
Just a few comments. Other look good.

File: enhancements/20230103-recurring-snapshot-cleanup.md (hunk 8 above)
Comment (innobead) [reply enhancements/20230103-recurring-snapshot-cleanup.md:58]: I think you just leverage the existing code, so the snapshot type doesn't matter.

File: enhancements/20230103-recurring-snapshot-cleanup.md (hunk 6 above)
Comment (c3y1huang) [reply enhancements/20230103-recurring-snapshot-cleanup.md]: > Not sure if retain is meaningful for snapshot-delete

I think you mean `snapshot-cleanup` right? I forgot to remove this, retain should not be effective with case `snapshot-cleanup`.

Comment (c3y1huang) [reply enhancements/20230103-recurring-snapshot-cleanup.md]: Removed, thanks!

File: enhancements/20230103-recurring-snapshot-cleanup.md (hunk 8 above)
Comment (c3y1huang) [reply enhancements/20230103-recurring-snapshot-cleanup.md:58]: Correct.

File: enhancements/20230103-recurring-snapshot-cleanup.md (hunk 9 above)
Comment (c3y1huang) [reply enhancements/20230103-recurring-snapshot-cleanup.md]: Sorry, this was written unclearly. I mean to propose mutating the `Recurringjob.Spec.Retain` to 0 when the task type is `snapshot-cleanup` since retain value has no effect on the purge.

Comment (innobead) [reply enhancements/20230103-recurring-snapshot-cleanup.md]: :100:

File: enhancements/20230103-recurring-snapshot-cleanup.md [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+https://github.com/longhorn/longhorn/issues/3836
+
+## Motivation
+
+### Goals
+
+Introduce new recurring job types:
+- `snapshot-delete`: periodically remove and purge all kinds of snapshots that exceed the retention count.
+- `snapshot-cleanup`: periodically purge to remove system snapshots.
```
Comment (innobead) [inline enhancements/20230103-recurring-snapshot-cleanup.md:18]:
```suggestion
- `snapshot-cleanup`: periodically purge removable or system snapshots.
```

**Review by innobead — CHANGES_REQUESTED** [review]

**Review by innobead — APPROVED** [review]
LGTM
